File: refactor_zoohub.py
import os
import re
from pathlib import Path

ZOOHUB_DIR = Path("d:/zoolearn(10.2.2026)/zoolearn_1st/src/components/zoohub")

def get_prefix(filename):
    name = Path(filename).stem
    return name[:4].lower()

def is_pseudo_or_vendor(name):
    pseudo = {'hover', 'active', 'focus', 'visited', 'first-child', 'last-child',
              'only-child', 'first-of-type', 'last-of-type', 'nth-child', 'nth-of-type',
              'before', 'after', 'placeholder', 'focus-within', 'root', 'not'}
    return name in pseudo or name.startswith(('webkit-', 'moz-', 'ms-'))

def already_has_prefix(name, prefix):
    return name.startswith(prefix + '-')

def extract_css_classes_ids(content):
    classes = set()
    ids = set()
    
    for match in re.finditer(r'\.([a-zA-Z_][\w-]*)', content):
        name = match.group(1)
        if not is_pseudo_or_vendor(name):
            classes.add(name)
    
    for match in re.finditer(r'#([a-zA-Z_][\w-]*)', content):
        name = match.group(1)
        if not re.match(r'^[0-9a-fA-F]{3,8}$', name):
            ids.add(name)
    
    return sorted(classes, key=len, reverse=True), sorted(ids, key=len, reverse=True)

def extract_jsx_classes_ids(content):
    classes = set()
    ids = set()
    
    for match in re.finditer(r'className\s*=\s*["\']([^"\']+)["\']', content):
        for cls in match.group(1).split():
            if cls:
                classes.add(cls)
    
    for match in re.finditer(r'\bid\s*=\s*["\']([^"\']+)["\']', content):
        ids.add(match.group(1))
    
    for match in re.finditer(r'querySelector(?:All)?\s*\(\s*["\']([^"\']+)["\']\s*\)', content):
        selector = match.group(1)
        if selector.startswith('.'):
            classes.add(selector[1:])
        elif selector.startswith('#'):
            ids.add(selector[1:])
    
    for match in re.finditer(r'getElementById\s*\(\s*["\']([^"\']+)["\']\s*\)', content):
        ids.add(match.group(1))
    
    return sorted(classes, key=len, reverse=True), sorted(ids, key=len, reverse=True)

def prefix_css(content, prefix, classes, ids):
    result = content
    
    for cls in classes:
        if not already_has_prefix(cls, prefix):
            pattern = r'\.' + re.escape(cls) + r'\b'
            replacement = '.' + prefix + '-' + cls
            result = re.sub(pattern, replacement, result)
    
    for id_name in ids:
        if not already_has_prefix(id_name, prefix):
            pattern = r'#' + re.escape(id_name) + r'\b'
            replacement = '#' + prefix + '-' + id_name
            result = re.sub(pattern, replacement, result)
    
    return result

def prefix_jsx(content, prefix, classes, ids):
    result = content
    
    # Process classes
    for cls in classes:
        if not already_has_prefix(cls, prefix):
            new_cls = prefix + '-' + cls
            
            # className attributes
            # Match whole className="..." with word boundary
            pattern1 = r'className="([^"]*\b)' + re.escape(cls) + r'\b([^"]*)"'
            result = re.sub(pattern1, r'className="\1' + new_cls + r'\2"', result)
            
            pattern2 = r"className='([^']*\b)" + re.escape(cls) + r"\b([^']*)'"
            result = re.sub(pattern2, r"className='\1" + new_cls + r"\2'", result)
            
            # querySelector
            pattern3 = r'querySelector\("' + re.escape('.' + cls) + r'"\)'
            result = re.sub(pattern3, 'querySelector(".' + new_cls + '")', result)
            
            pattern4 = r"querySelector\('" + re.escape('.' + cls) + r"'\)"
            result = re.sub(pattern4, "querySelector('." + new_cls + "')", result)
            
            pattern5 = r'querySelectorAll\("' + re.escape('.' + cls) + r'"\)'
            result = re.sub(pattern5, 'querySelectorAll(".' + new_cls + '")', result)
            
            pattern6 = r"querySelectorAll\('" + re.escape('.' + cls) + r"'\)"
            result = re.sub(pattern6, "querySelectorAll('." + new_cls + "')", result)
    
    # Process IDs
    for id_name in ids:
        if not already_has_prefix(id_name, prefix):
            new_id = prefix + '-' + id_name
            
            # id attributes
            pattern1 = r'\bid="' + re.escape(id_name) + r'"'
            result = re.sub(pattern1, 'id="' + new_id + '"', result)
            
            pattern2 = r"\bid='" + re.escape(id_name) + r"'"
            result = re.sub(pattern2, "id='" + new_id + "'", result)
            
            # querySelector with #
            pattern3 = r'querySelector\("' + re.escape('#' + id_name) + r'"\)'
            result = re.sub(pattern3, 'querySelector("#' + new_id + '")', result)
            
            pattern4 = r"querySelector\('" + re.escape('#' + id_name) + r"'\)"
            result = re.sub(pattern4, "querySelector('#" + new_id + "')", result)
            
            pattern5 = r'querySelectorAll\("' + re.escape('#' + id_name) + r'"\)'
            result = re.sub(pattern5, 'querySelectorAll("#' + new_id + '")', result)
            
            pattern6 = r"querySelectorAll\('" + re.escape('#' + id_name) + r"'\)"
            result = re.sub(pattern6, "querySelectorAll('#" + new_id + "')", result)
            
            # getElementById
            pattern7 = r'getElementById\("' + re.escape(id_name) + r'"\)'
            result = re.sub(pattern7, 'getElementById("' + new_id + '")' + result)
            
            pattern8 = r"getElementById\('" + re.escape(id_name) + r"'\)"
            result = re.sub(pattern8, "getElementById('" + new_id + "')", result)
    
    return result

def process_file(filepath):
    prefix = get_prefix(filepath.name)
    
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            original = f.read()
    except:
        return False
    
    if filepath.suffix == '.css':
        classes, ids = extract_css_classes_ids(original)
        updated = prefix_css(original, prefix, classes, ids) if (classes or ids) else original
    elif filepath.suffix in {'.jsx', '.js', '.tsx', '.ts', '.html'}:
        classes, ids = extract_jsx_classes_ids(original)
        updated = prefix_jsx(original, prefix, classes, ids) if (classes or ids) else original
    else:
        return False
    
    if updated != original:
        try:
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(updated)
            return True
        except:
            return False
    
    return False

# Main
print("Starting refactoring...")

files = []
for ext in ['*.css', '*.jsx', '*.js', '*.tsx', '*.ts']:
    files.extend(ZOOHUB_DIR.rglob(ext))

print(f"Found {len(files)} files")

changed = 0
for filepath in files:
    if process_file(filepath):
        rel_path = filepath.relative_to(ZOOHUB_DIR)
        print(f"Changed: {rel_path}")
        changed += 1

print(f"\nChanged {changed} of {len(files)} files")
print("Complete!")

File: refactor-zoohub.js
const fs = require('fs');
const path = require('path');

const ZOOHUB_DIR = path.join(__dirname, 'src', 'components', 'zoohub');

function getPrefix(filename) {
    const name = path.basename(filename, path.extname(filename));
    return name.substring(0, 4).toLowerCase();
}

function getAllFiles(dir, fileList = []) {
    const items = fs.readdirSync(dir, { withFileTypes: true });

    for (const item of items) {
        const fullPath = path.join(dir, item.name);
        if (item.isDirectory()) {
            getAllFiles(fullPath, fileList);
        } else if (/\.(css|jsx|js|tsx|ts|html)$/i.test(item.name)) {
            fileList.push(fullPath);
        }
    }

    return fileList;
}

function extractCssClassesIds(content) {
    const classes = new Set();
    const ids = new Set();

    // Extract classes
    const classRegex = /\.([a-zA-Z_][\w-]*)/g;
    let match;
    while ((match = classRegex.exec(content)) !== null) {
        const name = match[1];
        if (!isPseudoOrVendor(name)) {
            classes.add(name);
        }
    }

    // Extract IDs  
    const idRegex = /#([a-zA-Z_][\w-]*)/g;
    while ((match = idRegex.exec(content)) !== null) {
        const name = match[1];
        if (!/^[0-9a-fA-F]{3,8}$/.test(name)) {
            ids.add(name);
        }
    }

    return { classes: Array.from(classes), ids: Array.from(ids) };
}

function extractJsxClassesIds(content) {
    const classes = new Set();
    const ids = new Set();

    // className="..." or className={'...'}
    const classNameRegex = /className\s*=\s*["']([^"']+)["']/g;
    let match;
    while ((match = classNameRegex.exec(content)) !== null) {
        const classNames = match[1].split(/\s+/);
        classNames.forEach(cn => cn && classes.add(cn));
    }

    // id="..." or id={'...'}
    const idRegex = /\bid\s*=\s*["']([^"']+)["']/g;
    while ((match = idRegex.exec(content)) !== null) {
        ids.add(match[1]);
    }

    // querySelector
    const querySelectorRegex = /querySelector(?:All)?\s*\(\s*["']([^"']+)["']\s*\)/g;
    while ((match = querySelectorRegex.exec(content)) !== null) {
        const selector = match[1];
        if (selector.startsWith('.')) {
            classes.add(selector.substring(1));
        } else if (selector.startsWith('#')) {
            ids.add(selector.substring(1));
        }
    }

    // getElementById
    const getByIdRegex = /getElementById\s*\(\s*["']([^"']+)["']\s*\)/g;
    while ((match = getByIdRegex.exec(content)) !== null) {
        ids.add(match[1]);
    }

    return { classes: Array.from(classes), ids: Array.from(ids) };
}

function isPseudoOrVendor(name) {
    const pseudo = [
        'hover', 'active', 'focus', 'visited', 'first-child', 'last-child',
        'only-child', 'first-of-type', 'last-of-type', 'nth-child', 'nth-of-type',
        'before', 'after', 'placeholder', 'focus-within', 'root', 'not'
    ];
    return pseudo.includes(name) || name.startsWith('webkit-') || name.startsWith('moz-') || name.startsWith('ms-');
}

function alreadyHasPrefix(name, prefix) {
    return name.startsWith(prefix + '-');
}

function escapeRegex(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function prefixCss(content, prefix, classes, ids) {
    let result = content;

    const sortedClasses = classes.sort((a, b) => b.length - a.length);
    const sortedIds = ids.sort((a, b) => b.length - a.length);

    for (const className of sortedClasses) {
        if (!alreadyHasPrefix(className, prefix)) {
            const escaped = escapeRegex(className);
            const regex = new RegExp(`\\.${escaped}(?=[^\\w-]|$)`, 'g');
            result = result.replace(regex, `.${prefix}-${className}`);
        }
    }

    for (const id of sortedIds) {
        if (!alreadyHasPrefix(id, prefix)) {
            const escaped = escapeRegex(id);
            const regex = new RegExp(`#${escaped}(?=[^\\w-]|$)`, 'g');
            result = result.replace(regex, `#${prefix}-${id}`);
        }
    }

    return result;
}

function prefixJsx(content, prefix, classes, ids) {
    let result = content;

    const sortedClasses = classes.sort((a, b) => b.length - a.length);
    const sortedIds = ids.sort((a, b) => b.length - a.length);

    for (const className of sortedClasses) {
        if (!alreadyHasPrefix(className, prefix)) {
            const escaped = escapeRegex(className);

            // className attributes - handle space-separated classes
            result = result.replace(
                new RegExp(`(className\\s*=\\s*["'])([^"']*)`, 'g'),
                (fullMatch, before, classStr) => {
                    const updated = classStr.split(/\s+/).map(cls => {
                        return cls === className ? `${prefix}-${className}` : cls;
                    }).join(' ');
                    return before + updated;
                }
            );

            // querySelector with .
            result = result.replace(
                new RegExp(`(querySelector(?:All)?\\s*\\(\\s*["'])\\.${escaped}(["']\\s*\\))`, 'g'),
                `$1.${prefix}-${className}$2`
            );
        }
    }

    for (const id of sortedIds) {
        if (!alreadyHasPrefix(id, prefix)) {
            const escaped = escapeRegex(id);

            // id attributes
            result = result.replace(
                new RegExp(`(\\bid\\s*=\\s*["'])${escaped}(["'])`, 'g'),
                `$1${prefix}-${id}$2`
            );

            // querySelector  with #
            result = result.replace(
                new RegExp(`(querySelector(?:All)?\\s*\\(\\s*["'])#${escaped}(["']\\s*\\))`, 'g'),
                `$1#${prefix}-${id}$2`
            );

            // getElementById
            result = result.replace(
                new RegExp(`(getElementById\\s*\\(\\s*["'])${escaped}(["']\\s*\\))`, 'g'),
                `$1${prefix}-${id}$2`
            );
        }
    }

    return result;
}

function processFile(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const prefix = getPrefix(filePath);
    const originalContent = fs.readFileSync(filePath, 'utf8');
    let newContent = originalContent;

    if (ext === '.css') {
        const { classes, ids } = extractCssClassesIds(originalContent);
        if (classes.length > 0 || ids.length > 0) {
            newContent = prefixCss(originalContent, prefix, classes, ids);
        }
    } else if (['.jsx', '.js', '.tsx', '.ts', '.html'].includes(ext)) {
        const { classes, ids } = extractJsxClassesIds(originalContent);
        if (classes.length > 0 || ids.length > 0) {
            newContent = prefixJsx(originalContent, prefix, classes, ids);
        }
    }

    if (newContent !== originalContent) {
        fs.writeFileSync(filePath, newContent, 'utf8');
        return true;
    }

    return false;
}

// Main
console.log('🚀 Starting systematic refactoring...\n');

const allFiles = getAllFiles(ZOOHUB_DIR);
console.log(`Found ${allFiles.length} files\n`);

let changedCount = 0;

for (const filePath of allFiles) {
    const relativePath = path.relative(ZOOHUB_DIR, filePath);
    const changed = processFile(filePath);

    if (changed) {
        console.log(`✅ ${relativePath}`);
        changedCount++;
    }
}

console.log(`\n📊 Files changed: ${changedCount}/${allFiles.length}`);
console.log('✅ Complete!');
